### PR TITLE
feat: Family Resources Survey Report NI (DfC)

### LINKED ISSUE #1885: data-source-candidate: Family Resources Survey Report, Northern Ireland
## Data Source Evaluation: Family Resources Survey Report, Northern Ireland

**Source**: https://www.communities-ni.gov.uk/publications/family-resources-survey-report-2024-2025
**Publisher**: Department for Communities (DfC), Northern Ireland (NI supplement to DWP FRS)
**Format**: Seven Excel (.xlsx) chapter files
**Direct downloads (2024-25 edition)**:
- C1a: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c1a-confidence-intervals-2425.xlsx`
- C2: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c2-income-and-state-support-2425.xlsx`
- C3: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c3-tenure-2425.xlsx`
- C4: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c4-pensions-savings-and-investments-2425.xlsx`
- C5: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c5-carers-carees-and-disability-2425.xlsx`
- C6: `https://www.communities-ni.gov.uk/system/files/2026-05/frs-c6-food-security-and-food-bank-usage-2425.xlsx`
**Accessibility**: 4/5 — `communities-ni.gov.uk` file hosting returns HTTP 200, no Cloudflare/auth; consistent `frs-c[N]-[topic]-[YYMM].xlsx` naming
**Update frequency**: Annual; 2024-25 published May 2026
**Historical coverage**: Comparable 4-year time series from 2021/22 (break-in-series from integration of survey and admin data); earlier NI editions likely available on the DfC publication listing

### What the data contains

NI-specific household survey supplement to the UK-wide DWP Family Resources Survey. Six chapters:

- **C2 — Income & State Support**: Sources of household income (employment vs state support vs other); households in receipt of specific benefits; NI vs UK comparison; time series 2021/22–2024/25
- **C3 — Tenure**: Owner-occupied, mortgaged, social rented, private rented; NI breakdown
- **C4 — Pensions, Savings & Investments**: Pension participation rates, savings levels by type
- **C5 — Carers, Carees & Disability**: Disability prevalence, receipt of disability benefits, carer burden
- **C6 — Food Security & Food Bank Usage**: Household food security status (secure/marginal/low/very low), food bank usage rates — notable as official NI food bank data is rare

### Schema (C2 — example)

```
year, country (NI|UK), income_source, pct_total_income
year, state_support_type, pct_households_ni, pct_households_uk
```

### Integration pattern

Closest pattern: `nisra/wellbeing.py` or `nisra/labour_market.py` — multi-chapter annual household survey with NI-specific breakdowns. DfC file structure (chapter files, stable URL pattern) mirrors DfC housing statistics ODS pattern.

Place in: `src/bolster/data_sources/nisra/family_resources_survey.py` (or `dfc/family_resources_survey.py`)

### Cross-dataset correlation opportunities

1. `nisra/labour_market`: employment rate vs income sources (wages vs state support) — Chapter 2 aligns directly
2. `ni_house_prices` + housing bulletin: tenure trends (Chapter 3) vs house price affordability
3. `nisra/wellbeing`: subjective wellbeing scores vs food security (Chapter 6) — poverty indicator
4. `nisra/population`: disability rates (Chapter 5) vs population age structure
5. DWP UK-wide FRS: NI data in Chapter 2 cross-validates against DWP regional breakdown

### Derived insights possible

1. State support dependency rate: % households receiving benefits by year, NI vs UK comparison
2. Food insecurity trend: Chapter 6 time series as poverty proxy
3. Disability and carer burden: Chapter 5 cross-referenced with HSC workforce staffing levels

### Complexity assessment

**Low-Medium** — 7 annual Excel files from stable URLs; each chapter has a small number of tables with consistent structure. Scraping the edition page for chapter links is the main complexity. 4-year time series is short but will lengthen annually.

**Recommendation**: RECOMMENDED — Chapter 6 (food security/food bank usage) is particularly novel; no existing bolster module covers this domain.

### Next steps for data-build agent

1. Scrape `https://www.communities-ni.gov.uk/publications?keywords=family+resources+survey+report` to discover latest edition URL
2. Scrape edition page for all chapter `.xlsx` download links
3. Implement per-chapter getters: `get_income_data()`, `get_tenure_data()`, `get_disability_data()`, `get_food_security_data()`
4. Key test: Chapter 2 — ~69% NI households receiving some state support (2024/25); Chapter 3 — tenure breakdown sums to ~100%
5. Add CLI `bolster nisra family-resources-survey --chapter [income|tenure|pensions|disability|food]`

### PR DESCRIPTION
**Pull request by andrewbolster**
## Summary

Adds `bolster.data_sources.dfc.family_resources_survey`, wrapping the Department for Communities' annual Family Resources Survey Report for Northern Ireland — household income and state support, housing tenure, carers and disability, and household food security.

- Runtime edition discovery walks the publication URL template backwards from the current year until it finds a page linking the `frs-c*.xlsx` workbooks, so new annual editions are picked up without a code change. Machine-readable tables begin with the 2024/25 edition.
- 14 accessors across the six published chapters, returning tidy frames with `sample_size` retained from each table's "Sample size (=100%)" footer row.
- Publication suppression markers are normalised: `".."` → `NaN`, `"-"` → `0.0`.
- New `bolster dfc frs` CLI command exposing every table via `--dataset`, with `--list-datasets`, `--summary`, `--save` and `--format table|csv|json`.

## Example insights

**Northern Ireland is the most food-secure UK country.** 94% of NI households are food secure vs 91% UK-wide; Inner London is the weakest region at 85%.

**Food insecurity tracks income-related benefit receipt, not benefit receipt generally.** Households on any income-related benefit are food secure only 82% of the time, against 93% for those on non-income-related benefits and 97% for households receiving no state support at all.

**Owner-occupation varies by 23 percentage points across councils.** Belfast City is lowest at 55%, against 78% in both Lisburn & Castlereagh and Ards & North Down — the same ordering the NIMDM deprivation ranks imply.

## Cross-validation

District disability prevalence was checked against mean NIMDM 2017 `health_disability_rank` per LGD. The correlation is negative as expected (Spearman −0.24; lower rank = more deprived), with Derry City & Strabane both the highest-prevalence district (34%) and the most health-deprived. The correlation is weak rather than strong, which is consistent with FRS district sample sizes of 358–986 and a 2017 deprivation vintage.

## Usage

```python
from bolster.data_sources.dfc import family_resources_survey as frs

edition, url = frs.find_latest_edition()   # ('2024/25', 'https://www.communities-ni.gov.uk/...')

frs.get_food_security_by_region()          # UK countries + English regions
frs.get_food_security_by_state_support()   # by benefit receipt
frs.get_tenure_by_district()               # 11 LGDs
frs.get_disability_by_district()
frs.get_income_sources()                   # NI vs UK, by source
```

```console
$ bolster dfc frs --list-datasets
$ bolster dfc frs --dataset food-security-region
$ bolster dfc frs --dataset tenure-district --summary
$ bolster dfc frs --dataset income-sources --format csv --save income.csv
```

## Test plan

- [x] 121 integrity tests pass against live DfC workbooks (`tests/test_dfc_family_resources_survey_integrity.py`)
- [x] 18 module doctests pass
- [x] Module line coverage 93.75%
- [x] Full suite: 2186 passed, 7 skipped (31 errors are a pre-existing upstream HTTP failure in `test_psni_police_ombudsman_integrity.py`, unrelated to this branch)
- [x] `pre-commit run --all-files` clean
- [x] CLI exercised manually for `--list-datasets`, `--summary` and table output

Closes #1885

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by andrewbolster — COMMENTED** [review]
## data-review: PR #2026 — Family Resources Survey Report NI (DfC)

Reviewed against the `data-review` checklist in AGENTS.md. Originating issue: #1885 (`data-source-candidate`, evaluated RECOMMENDED, correctly referenced via `Closes #1885`).

### CI Status — ⚠️ needs attention before merge

`gh pr checks`-equivalent via the API shows **`build (3.11)` failed** (`build (3.12)`/`build (3.13)` cancelled as a consequence). Root-caused via `get_job_logs`:

```
FAILED tests/test_nisra_labour_market_integrity.py::TestLGDEmploymentIntegrity::test_employment_components_sum
  - AssertionError: Armagh City, Banbridge and Craigavon: Full-time (90) + Part-time (23) = 113 != Total employment (115)
16 errors in tests/test_health_ni_disease_prevalence_gp_practice_integrity.py
  - NISRADataNotFoundError: No publication matching 'disease prevalence' found on health-ni.gov.uk
= 1 failed, 2265 passed, 8 skipped, 204 warnings, 16 errors in 518.22s =
```

Neither failure touches `dfc/` — both are pre-existing, unrelated to this diff (labour market data revision + a health-ni.gov.uk publication-search miss), confirming `changed_files` for this PR is only README.md, docs/data_sources.rst, cli.py, dfc/__init__.py, dfc/family_resources_survey.py, and the new test file.

However, **note the PR description's test-plan claims "31 errors ... `test_psni_police_ombudsman_integrity.py`, unrelated to this branch"** — that doesn't match this run at all (this run has 16 errors in `health_ni` GP-practice tests, not ombudsman). Worth re-verifying the failure signature is still the accepted "known flake" before merging, since it's drifted since the PR description was written.

**More importantly: `mergeable_state` is `"dirty"` — this PR cannot currently be merged**, and the cause is concrete, not cosmetic:

- `src/bolster/data_sources/dfc/__init__.py` (this PR) is an entirely new file (`@@ -0,0 +1,15 @@`) exporting only `family_resources_survey`. But **main already has this file**, added by the since-merged Child Maintenance Service PR, exporting `child_maintenance`. Git can't reconcile two independent "new" versions of the same file.
- `src/bolster/cli.py` (this PR) adds `@cli.group() def dfc(): ...` as a brand-new group after `daera_waste_cmd`. **Main's `cli.py` already defines `def dfc():` at line 7460**, with a `@dfc.command(name="child-maintenance")` registered under it. Merging as-is will either hard-conflict or (if somehow auto-resolved) silently orphan the `bolster dfc child-maintenance` command, since Click's `Group.add_command` overwrites same-named commands without raising.

This branch needs a rebase onto current `main` with `dfc/__init__.py` updated to export **both** `child_maintenance` and `family_resources_survey`, and the `dfc_frs_cmd` added under the **existing** `dfc` group in `cli.py` rather than redefining it. README.md's coverage-table insertion point (after "Translink Live...") sits directly above main's existing "Child Maintenance Service Statistics" row and will likely need the same manual reconciliation, even though it may merge silently.

### Code Quality

- Type hints and docstrings (Args/Returns/Example) are consistently present and good quality throughout `family_resources_survey.py`.
- Uses `bolster.utils.web.session` and `bolster.utils.cache.CachedDownloader` correctly for the workbook download itself (`family_resources_survey.py:89`, `:184`).
- **Network calls in `find_latest_edition()` (`family_resources_survey.py:118`) and `get_chapter_urls()` (`:148-149`) are not wrapped in try/except**, unlike every comparable module in the codebase:
  - The sibling module in the same package, `dfc/child_maintenance.py::list_publications()`, wraps its `session.get()` in try/except and raises `CMSDataNotFoundError` on failure.
  - `nisra/_base.py::find_publication_link()` does the same for `NISRADataNotFoundError`.
  A transient network error (timeout, DNS blip, connection reset) during edition discovery will surface as a raw `requests` exception rather than `FRSDataError`/`FRSDataNotFoundError`, which is inconsistent with the rest of the codebase's error-handling contract and with the CLI's `except Exception` catch-all being the only thing standing between a raw traceback and the user (`cli.py`'s `dfc_frs_cmd`, which does catch broadly, so end-user impact is minor — but callers of the module functions directly get an undocumented exception type).
- **No `validate_data()` function**, unlike the module template in AGENTS.md and unlike the sibling `dfc/child_maintenance.py::validate_data()` (which checks required columns, row-count floor, known enums, value ranges). There's also no `FRSValidationError` in the exception hierarchy (`family_resources_survey.py:92-96` only defines `FRSDataError`/`FRSDataNotFoundError`). Given 14 accessors each returning differently-shaped frames, a single `validate_data()` may not be practical, but at minimum this is a documented deviation from both the team template and the established pattern in the same subpackage — worth a one-line note in the module docstring if intentional.
- Minor: `CHAPTERS` (`family_resources_survey.py:74-81`) declares 6 chapters (`confidence_intervals`, `income`, `tenure`, `pensions`, `carers_disability`, `food_security`), but only 4 are ever passed to `_read_table` by a public accessor — `confidence_intervals` and `pensions` are resolved by `get_chapter_urls()`/`list_chapters()` but have no accessor function. Not wrong (module docstring says "covers four of them"), but `list_chapters()` advertises two chapters a caller can't actually get data from.

### Testing

- Excellent structure: `@pytest.mark.network` + `scope="class"` fixtures for the real-data integrity suite, plus a separate no-network `TestParsingHelpers` class unit-testing `_clean_value`, `_label`, `_is_footer`, `_find_header_row`, `_data_rows`, `_sample_sizes`, `_composition_sections`, `_normalise_year` — this substitutes reasonably for the missing `validate_data()` edge-case tests called for in AGENTS.md's `TestValidation` template, since there's no `validate_data()` to test against empty/malformed frames.
- No mocks; real workbooks. Good coverage of both the tidy-output assertions (percentages sum to ~100, sample sizes positive, cross-dataset plausibility like "NI more food secure than UK") and structural edge cases.
- Per the PR description, 121 tests / 93.75% module coverage — consistent with the depth visible in the diff.

### Integration

- CLI: `bolster dfc frs` added with `--dataset`, `--list-datasets`, `--summary`, `--save`, `--format`, following the same pattern as `daera_waste_cmd` immediately above it in the diff — good consistency in isolation, but see the merge-conflict note above; it must land under the pre-existing `dfc` group, not a duplicate one.
- README.md and `docs/data_sources.rst` both updated with the new module — good, though also subject to the rebase noted above.
- `Closes #1885` present and correct.

### Opportunities (shared utilities)

This is now the **second** Excel-scraping module in `dfc/` (after `child_maintenance.py`), and the two independently reimplement a near-identical low-level toolkit:

| Concern | `family_resources_survey.py` | `child_maintenance.py` |
|---|---|---|
| Cell → float, honouring suppression markers | `_clean_value` | `_parse_value` |
| Cell → trimmed label | `_label` | `_clean_label` |
| Locate header row by first-column label | `_find_header_row` | `_find_header_row` (same name, separate implementation) |
| End-of-data-block detection | `_is_footer` | `_is_source_row` |
| Wrapped, cached download with domain exception | inline `_downloader.download()` (uncaught `DownloadError`) | `download_file()` wrapper |

Given `nisra/_base.py` already exists precisely to hold this kind of cross-module Excel/scraping helper (`find_header_row`, `safe_float`, `scrape_download_links`, `find_publication_link`, `make_absolute_url`), and `dfc/` now has two modules reinventing the same primitives independently (with `_find_header_row` literally colliding in name but not implementation), this looks like a solid candidate for a `dfc/_base.py` (mirroring `nisra/_base.py`) — or, since the "locate header row / strip suppression markers / detect footer row" pattern is generic to any suppression-marked government stats workbook, a source-agnostic `utils/excel.py`. Worth flagging to `data-maintenance` if it isn't already tracked — this is exactly the "repeated pattern across ≥2 PRs" trigger in that agent's workflow.

Secondary, lower-priority observation: `find_latest_edition()` probes the dated URL pattern with an uncached `session.get()` on every call, and every one of the 14 public accessors re-runs both `find_latest_edition()` and `get_chapter_urls()` from scratch (neither result is memoized) before touching the cached workbook download. Across a full test run or a script calling several accessors, that's several uncached HTML GETs per accessor. `child_maintenance.py`'s `list_publications()` has the same characteristic, so this isn't new to this PR, but a small `functools.lru_cache`-style memo (with a `force_refresh` bypass, matching the existing kwarg convention) on edition/chapter-URL discovery would reduce load on `communities-ni.gov.uk` and speed up the test suite.

---
_Generated by [Claude Code](https://claude.ai/code)_
